### ISSUE TITLE
promisified functions are resolved right after execution

### ISSUE THREAD

**Issue by tomgrossman** [Automattic/mongoose#4246]
I think this bug is related to this feature from the latest version:

**feat(model): emit error event when cb throws exception**
[https://github.com/Automattic/mongoose/commit/36a47c8a373dae59e0d105ecc28b8c40d30cf2ae](url)

I am using bluebird's promisifyAll on mongoose model.
I noticed that when executing an "Async" function, the result is resolved right after the execution, without waiting to finish.

For example, with version **4.4.20** I used the 
`Model.ensureIndexesAsync()`
function, and the promise result was resolved only after the indexing was done.

But now, with version **4.5.0**, the same execution on the same collection takes no time at all, and the function is resolved while the indexing is still happening.

I saw in the code that you wrap the callback with "try catch", can it be that it affects the execution of promisified functions?

**Comment by vkarpov15** [comment]
Hmm it shouldn't. I'm curious though, why are you using bluebirds promisification if you can just make mongoose return bluebird promises?

**Comment by tomgrossman** [comment]
@vkarpov15 Good question, I think is because there isn't enough documentation about it and how to use it, and what is supported and what is not.
I couldn't find some information like when do I need to use ` exec()` and when I don't need to.

But after the last 2 issues I've opened, I'm thinking more about using it and not promisify.

**Comment by vkarpov15** [comment]
Take a look at http://mongoosejs.com/docs/promises.html, should be helpful. In the meantime, can you show me what your promisification code looks like? Would be handy to investigate this.

**Comment by tomgrossman** [comment]
Well, so I've converted all my code to use mongoose promises with bluebird now as you suggested.

Anyway, I did some further investigation on this issue and it seems like there is no issue with the promises (both promisifyAll and mongoose promises).

Like I wrote in the original issue, I had a problem with the `ensureIndexes()` function which returned instantly on a large collection.

So there can be 2 options: there is a bug in this function, or there is no bug at all.

My schema has autoIndex enabled:
`mySchema.set('autoIndex', true);`

The function works as follows (with mongoose promises with bluebird):

```javascript
return myModel.collection.drop().then(
    function () {
        return myModel.fillSomething();
    }
).then(
    console.log('before indexing');
    return myModel.ensureIndexes();
).then(
    function () {
        console.log('after indexing');
        return myModel.doSomethin(); // should be done only after indexing is done!!
    }
)
```

during the execution of this code, some other nodes may be starting so the `autoIndex` goes into action.
it takes ~5 minutes to complete the indexing process of this collection after it fills with data.
So what happens is, that if the indexing has already started (even if it's just at the beginning), the `ensureIndexes()` function returns instantly instead of waiting for the indexing process to finish.

So the question is if it works as expected or the `ensureIndexes()` should be fixed.
